lib/socket.c

```diff
@@ -567,100 +567,111 @@ iscsi_read_from_socket(struct iscsi_context *iscsi)
 {
 	struct iscsi_in_pdu *in;
 	ssize_t data_size, count, padding_size;
+	int available;
 
-	if (iscsi->incoming == NULL) {
-		iscsi->incoming = iscsi_szmalloc(iscsi, sizeof(struct iscsi_in_pdu));
-		iscsi->incoming->hdr = iscsi_szmalloc(iscsi, ISCSI_RAW_HEADER_SIZE + ISCSI_DIGEST_SIZE);
-		if (iscsi->incoming == NULL) {
-			iscsi_set_error(iscsi, "Out-of-memory: failed to malloc iscsi_in_pdu");
-			return -1;
-		}
+	/* check how much data is in the input buffer of the socket and read as many
+	 * PDUs as available */
+	if (ioctl(iscsi->fd, FIONREAD, &available) < 0) {
+		return -1;
 	}
-	in = iscsi->incoming;
 
-	/* first we must read the header, including any digests */
-	if (in->hdr_pos < ISCSI_HEADER_SIZE) {
-		/* try to only read the header, the socket is nonblocking, so
-		 * no need to limit the read to what is available in the socket
-		 */
-		count = ISCSI_HEADER_SIZE - in->hdr_pos;
-		count = recv(iscsi->fd, &in->hdr[in->hdr_pos], count, 0);
-		if (count == 0) {
-			return -1;
+	while (available > 0) {
+		if (iscsi->incoming == NULL) {
+			iscsi->incoming = iscsi_szmalloc(iscsi, sizeof(struct iscsi_in_pdu));
+			iscsi->incoming->hdr = iscsi_szmalloc(iscsi, ISCSI_RAW_HEADER_SIZE + ISCSI_DIGEST_SIZE);
+			if (iscsi->incoming == NULL) {
+				iscsi_set_error(iscsi, "Out-of-memory: failed to malloc iscsi_in_pdu");
+				return -1;
+			}
 		}
-		if (count < 0) {
-			if (errno == EINTR || errno == EAGAIN) {
-				return 0;
+		in = iscsi->incoming;
+
+		/* first we must read the header, including any digests */
+		if (in->hdr_pos < ISCSI_HEADER_SIZE) {
+			/* try to only read the header, the socket is nonblocking, so
+			 * no need to limit the read to what is available in the socket
+			 */
+			count = ISCSI_HEADER_SIZE - in->hdr_pos;
+			count = recv(iscsi->fd, &in->hdr[in->hdr_pos], count, 0);
+			if (count == 0) {
+				/* remote side has closed the socket. */
+				return -1;
 			}
-			iscsi_set_error(iscsi, "read from socket failed, "
-				"errno:%d", errno);
+			if (count < 0) {
+				if (errno == EINTR || errno == EAGAIN) {
+					break;
+				}
+				iscsi_set_error(iscsi, "read from socket failed, "
+					"errno:%d", errno);
+				return -1;
+			}
+			in->hdr_pos  += count;
+			available -= count;
```

Comment:
Would need a check if available <= 0  here and abort.

Otherwise, if we have only reveiced exacly just the header but not the payload, we would still have available > 0 here
but as there is no more data to receive, lines 636 or 648 would return 0 and thus 650 would incorrectly think the socket has been clsoed.